iohub/fov.py

```diff
@@ -0,0 +1,215 @@
+
+    @abstractmethod
+    def __getitem__(
+        self,
+        key: Union[int, slice, tuple[Union[int, slice], ...]],
+    ) -> ArrayLike:
+        """
+        Returned object must support the ``__array__`` interface,
+        so that ``np.asarray(...)`` will work.
+        """
+        raise NotImplementedError
+
+    @property
+    def ndim(self) -> int:
+        return 5
+
+    @property
+    @abstractmethod
+    def dtype(self) -> np.dtype:
+        raise NotImplementedError
+
+    @property
+    def zyx_scale(self) -> tuple[float, float, float]:
+        """Helper function for FOV spatial scale."""
+        raise NotImplementedError
+
+    def __eq__(self, other: BaseFOV) -> bool:
+        if not isinstance(other, BaseFOV):
+            return False
+        return self.root.absolute() == other.root.absolute()
+
+
+class BaseFOVMapping(Mapping):
+    @abstractmethod
+    def __enter__(self) -> BaseFOVMapping:
+        """Open the underlying file and return self."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def __exit__(
+        self,
+        exc_type: Optional[Type[BaseException]],
+        exc_val: Optional[BaseException],
+        exc_tb: Optional[TracebackType],
+    ) -> bool:
+        """Close the files."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def __contains__(self, position_key: str) -> bool:
+        """Check if a position is present in the collection."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def __len__(self) -> int:
+        raise NotImplementedError
+
+    @abstractmethod
+    def __getitem__(self, position_key: str) -> BaseFOV:
+        """FOV key position to FOV object."""
+        raise NotImplementedError
+
+    @abstractmethod
+    def __iter__(self) -> Iterable[tuple[str, BaseFOV]]:
+        """Iterates over pairs of keys and FOVs."""
+        raise NotImplementedError
+
+
+class FOVDict(BaseFOVMapping):
+    """
+    Basic implementation of a mapping of strings to BaseFOVs.
+    """
+
+    def __init__(
+        self,
+        data_dict: Optional[dict[str, BaseFOV]] = None,
+        **kwargs,
+    ) -> None:
+        super().__init__()
+        self._data = {}
+
+        if data_dict is not None:
+            for key, fov in data_dict.items():
+                self._safe_insert(key, fov)
+
+        for key, fov in kwargs.items():
+            self._safe_insert(key, fov)
+
+    def _safe_insert(self, key: str, value: BaseFOV) -> None:
+        """Checks if types are correct and key is unique."""
+        if not isinstance(key, str):
+            raise TypeError(
+                f"{self.__class__.__name__} key must be str. "
+                f"Found {key} with type {type(key)}"
+            )
+
+        if not isinstance(value, BaseFOV):
+            raise TypeError(
+                f"{self.__class__.__name__} value must subclass BaseFOV. "
+                f"Found {key} with value type {type(value)}"
+            )
+
+        if key in self:
```

Comment:
Is this redundant if `__init__` already receives a `dict`?

Reply:
It can be initialized in two different ways, so this additional check is necessary, and we only pay a very small overhead.

- kwargs init: `FOVDict(green=..., red=...)`
- dict init: `FOVDict({"green": ..., "red": ...})`
- both: `FOVDict({"green": ...}, red=...)`